flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2256,6 +2310,38 @@ class ArrayExprLowering {
     return {iters, insPt};
   }
 
+  fir::ArrayLoadOp
+  createAndLoadSomeArrayTemp(mlir::Type type,
+                             llvm::ArrayRef<mlir::Value> shape) {
+    auto seqTy = type.dyn_cast<fir::SequenceType>();
+    assert(seqTy && "must be an array");
+    auto loc = getLoc();
+    // TODO: Need to thread the length parameters here. For character, they may
+    // differ from the operands length (e.g concatenation). So the array loads
+    // type parameters are not enough.
+    if (auto charTy = seqTy.getEleTy().dyn_cast<fir::CharacterType>())
+      if (charTy.hasDynamicLen())
+        TODO(loc, "character array expression temp with dynamic length");
```

Comment:
Should be on its way...

Reply:
Mmh, that still require some work after your patch. The length of the result may not be the one of the operands (e.g. concatenation). So I cannot just take the arrayLoad one here. I probably need a finer grain control over the length.